cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
   }
 };
 
+// ----------------------------------------------------------------------
+// String to Timestamp
+
+template <>
+struct CastFunctor<TimestampType, StringType> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    //     const auto& out_type = checked_cast<const TimestampType&>(*output->type);
+    using out_type = TimestampType::c_type;
+
+    StringArray input_array(input.Copy());
+    auto out_data = output->GetMutableValues<out_type>(1);
+    internal::StringConverter<TimestampType> converter(output->type);
+
+    for (int64_t i = 0; i < input.length; ++i, ++out_data) {
+      if (input_array.IsNull(i)) {
+        continue;
```

Comment:
Should you push a null to out_data?

Reply:
I think nulls are simply copied as a separate pass.